Instrument obsreport.receiver metrics with otel go

paivagustavo opened the pull request:
**Description:** Instruments existing `obsreport.Receiver` metrics  with otel-go, side-by-side with the existing opencensus metrics.


This uses the same `UseOtelForInternalMetricsfeatureGateID` feature gate, adding a `UseOtelForMetrics()` method on the `obsreportconfig` that should be used by components of the collector core to check what sdk should be used for instrumentation.

**Link to tracking Issue:** part of https://github.com/open-telemetry/opentelemetry-collector/issues/816

**Testing:** Manual tests comparing the current exported metrics and the metrics from the otel-go sdk after enabling the feature gate `--feature-gates=telemetry.useOtelForInternalMetrics`

bogdandrutu commented on internal/obsreportconfig/obsmetrics/obs_receiver.go:
```diff
@@ -83,4 +87,40 @@ var (
 		"Number of log records that could not be pushed into the pipeline.",
 		stats.UnitDimensionless)
+
+	otelMeter     = global.Meter(MeterName)
+	AttrReceiver  = attribute.Key(ReceiverKey)
+	AttrTransport = attribute.Key(TransportKey)
```
Why using the global meter?

paivagustavo replied:
I've moved those to be created inside of `obsreport.Receiver` using the `ReceiverCreateSettings.MeterProvider`.

bogdandrutu commented on internal/obsreportconfig/obsmetrics/obs_receiver.go:
```diff
@@ -83,4 +87,40 @@ var (
+	AttrReceiver  = attribute.Key(ReceiverKey)
+	AttrTransport = attribute.Key(TransportKey)
+
+	// Receiver metrics created via the OpenTelemetry Go APIs.
+	OtelReceiverAcceptedSpans, _ = otelMeter.SyncInt64().Counter(
+		ReceiverPrefix+AcceptedSpansKey,
+		instrument.WithDescription("Number of spans successfully pushed into the pipeline."),
+		instrument.WithUnit(unit.Dimensionless),
+	)
+	OtelReceiverRefusedSpans, _ = otelMeter.SyncInt64().Counter(
+		ReceiverPrefix+RefusedSpansKey,
+		instrument.WithDescription("Number of spans that could not be pushed into the pipeline."),
+		instrument.WithUnit(unit.Dimensionless),
+	)
+	OtelReceiverAcceptedMetricPoints, _ = otelMeter.SyncInt64().Counter(
+		ReceiverPrefix+AcceptedMetricPointsKey,
+		instrument.WithDescription("Number of metric points successfully pushed into the pipeline."),
+		instrument.WithUnit(unit.Dimensionless),
+	)
+	OtelReceiverRefusedMetricPoints, _ = otelMeter.SyncInt64().Counter(
+		ReceiverPrefix+RefusedMetricPointsKey,
+		instrument.WithDescription("Number of metric points that could not be pushed into the pipeline."),
+		instrument.WithUnit(unit.Dimensionless),
+	)
+	OtelReceiverAcceptedLogRecords, _ = otelMeter.SyncInt64().Counter(
+		ReceiverPrefix+AcceptedLogRecordsKey,
+		instrument.WithDescription("Number of log records successfully pushed into the pipeline."),
+		instrument.WithUnit(unit.Dimensionless),
+	)
+	OtelReceiverRefusedLogRecords, _ = otelMeter.SyncInt64().Counter(
+		ReceiverPrefix+RefusedLogRecordsKey,
+		instrument.WithDescription("Number of log records that could not be pushed into the pipeline."),
+		instrument.WithUnit(unit.Dimensionless),
+	)
```
these can be defined in the obsreport package now, since we don't need to share them between other internal packages.

bogdandrutu commented on obsreport/obsreport_receiver.go:
```diff
@@ -163,23 +175,42 @@ func (rec *Receiver) endOp(
-		case config.LogsDataType:
-			acceptedMeasure = obsmetrics.ReceiverAcceptedLogRecords
-			refusedMeasure = obsmetrics.ReceiverRefusedLogRecords
+		if rec.otelEnabled {
+			var acceptedMeasure, refusedMeasure syncint64.Counter
+			switch dataType {
+			case config.TracesDataType:
+				acceptedMeasure = obsmetrics.OtelReceiverAcceptedSpans
+				refusedMeasure = obsmetrics.OtelReceiverRefusedSpans
+			case config.MetricsDataType:
+				acceptedMeasure = obsmetrics.OtelReceiverAcceptedMetricPoints
+				refusedMeasure = obsmetrics.OtelReceiverRefusedMetricPoints
+			case config.LogsDataType:
+				acceptedMeasure = obsmetrics.OtelReceiverAcceptedLogRecords
+				refusedMeasure = obsmetrics.OtelReceiverRefusedLogRecords
```
extract separate funcs "recordWithOtel" and "recordWithOC" to help with readability.

paivagustavo replied:
done.

bogdandrutu commented on obsreport/views.go:
```diff
@@ -0,0 +1,54 @@
+// AddViews to the OpenTelemetry Go SDK.
+// Views are only configurable at the SDK initialization, for this reason
+// components must call this function inside an init() function to be correctly initialized.
+func AddViews(v ...otelview.View) {
```
Why do you want to expose this functionality? What is the purpose of it?

bogdandrutu replied:
This can/should be added later. There are other nicer alternatives like allowing Factories to implement additional interfaces for this, or doing it in the NewFactory helper as optionals, etc.

paivagustavo replied:
I'll just remove this for now as this was just here as a preparation step migrations of some other components, thanks for sharing some nicer options as well!

bogdandrutu reviewed (commented):
We need this https://github.com/open-telemetry/opentelemetry-collector/pull/5939, which is to not break the metrics names at this time. We will at one point later revisit all names.

jaronoff97 commented on obsreport/obsreport_receiver.go:
```diff
@@ -63,7 +87,70 @@ func NewReceiver(cfg ReceiverSettings) *Receiver {
+		}
 	}
+
+	rec.acceptedSpansCounter, err = rec.meter.SyncInt64().Counter(
```
thought: I wonder if these should exist in a list/map that are then just accessed as needed. i.e.

```
        telemetryMap = map[config.DataType]map[string]*syncint64.Counter
        // iterate through [DataType] x ["accepted", "refused"] to initialize
...
	telemetryMap[dataType]["accepted"].Add(receiverCtx, int64(numAccepted), rec.otelAttrs...)
	telemetryMap[dataType]["refused"].Add(receiverCtx, int64(numRefused), rec.otelAttrs...)
```

bogdandrutu replied:
Accessing a map is slower than direct access. So I prefer this way.

jaronoff97 reviewed (approved):
Just one thought, otherwise looks great!

paivagustavo commented:
> We need this https://github.com/open-telemetry/opentelemetry-collector/pull/5939, which is to not break the metrics names at this time. We will at one point later revisit all names.

@bogdandrutu  this pr actually fixes the name inconsistencies wrapping the `prometheus.Registry`, this might be easier than trying add `otelcol_` prefix to all metrics (including contrib metrics):
```
prometheus.WrapRegistererWithPrefix("otelcol_", prometheus.WrapRegistererWith(promLabels, registry))
```

If we still wants to move on with #5939, I can create a copy of `obsmetrics.ReceiverPrefix`, so this won't make `otelcol_otelcol_` metrics and fix this after that get merged.

bogdandrutu commented on internal/obsreportconfig/obsmetrics/obsmetrics.go:
```diff
@@ -19,4 +19,10 @@ package obsmetrics // import "go.opentelemetry.io/collector/internal/obsreportco
 const (
 	NameSep = "/"
+
+	MeterName = "io.opentelemetry.collector"
```
Why this name? Do we have the same name for receivers/exporters? How does this affect the metrics exposed to Prometheus? /cc @dashpole 

In order to be a no-op change between OC -> Otel for the moment, I think we should have an empty name for the moment. Then when we have the final list of the metrics, we should agree on what scope name we use.

paivagustavo replied:
This can be removed since I'm no longer using the global meter and using the `ReceiverCreateSettings.MeterProvider` instead, there I'm using the same name as the trace name which is `cfg.ReceiverID.String()`.


> In order to be a no-op change between OC -> Otel for the moment, I think we should have an empty name for the moment.

I think this does not matter for the prometheus exporter,  as the otel scope is ignored now and will be added as additional attributes and metrics per https://github.com/open-telemetry/opentelemetry-specification/pull/2703

bogdandrutu commented on obsreport/obsreport.go:
```diff
@@ -17,10 +17,17 @@ package obsreport // import "go.opentelemetry.io/collector/obsreport"
 	}
 }
+
+func InstrumentWithOtel() bool {
```
Why public? Who is the user of this?

paivagustavo replied:
I planned this to be public for all components that want to double instrument with otel-go, but as I've discovered the  `TelemetrySettings` struct, that seems to be a more suitable place, what do you think about that?

bogdandrutu commented on service/telemetry.go:
```diff
@@ -213,15 +221,39 @@ func (tel *telemetryInitializer) initOpenCensus(cfg telemetry.Config, telAttrs m
+		promLabels[sanitizePrometheusKey(k)] = v
+	}
+
+	wrappedRegisterer := prometheus.WrapRegistererWithPrefix("otelcol_", prometheus.WrapRegistererWith(promLabels, registry))
+	if err := wrappedRegisterer.Register(exporter.Collector); err != nil {
```
Let's split this change in a separate PR.

paivagustavo replied:
Created https://github.com/open-telemetry/opentelemetry-collector/pull/6223 for this.

jpkrohling reviewed (commented):
This looks sane to me,  a review from @Aneurysm9 would be helpful though.

bogdandrutu commented on component/telemetry.go:
```diff
@@ -33,6 +33,10 @@ type TelemetrySettings struct {
 	// MeterProvider that the factory can pass to other instrumented third-party libraries.
 	MeterProvider metric.MeterProvider
 
+	// UseOtelForMetrics controls whether the collector should use otel go sdk for
+	// internal telemetry metrics.
+	UseOtelForMetrics bool
```
Why exposing this?

paivagustavo replied:
I've moved it to the internal package `obsreportconfig` so we can use it to the collector core component with otel.

bogdandrutu reviewed (changes requested):
Current PR exposes new temporary public APIs because the OC bridge is not ready and we cannot really start the proper migration to otel. 

I am not OK doing that.

paivagustavo commented:
@bogdandrutu  are you concerned about having this public `TelemetrySettings.UseOtelForMetrics`? Would it be better if we get rid of that and just started to record values to both sdks at the same time? Something like this:
```go
rec.recordWithOtel(receiverCtx, dataType, numAccepted, numRefused)
rec.recordWithOC(receiverCtx, dataType, numAccepted, numRefused)
```

I'm trying to understand what other paths we might have other than just waiting for the bridge to get ready.

bogdandrutu commented:
@paivagustavo to give you alternatives, I need to understand the goal :)

codeboten commented on internal/obsreportconfig/obsreportconfig.go:
```diff
@@ -20,9 +20,21 @@ import (
+	UseOtelForInternalMetricsfeatureGateID = "telemetry.useOtelForInternalMetrics"
+)
+
+// UseOtelForMetrics controls whether the Collector should use Otel for Collector's internal instrumentation.
```
nit: this method doesn't control anything but reports it 

```suggestion
// UseOtelForMetrics returns whether the Collector should use Otel for Collector's internal instrumentation.
```

Aneurysm9 replied:
I don't think this function is even needed since the feature gate ID constant is also exported.  It's easy enough for consumers to check the gate themselves and also hopefully avoids hiding the fact that a gate is being checked so it isn't called in a tight loop accidentally.

paivagustavo replied:
Just removed it.

codeboten commented on obsreport/obsreport_receiver.go:
```diff
@@ -36,6 +42,19 @@ type Receiver struct {
+	meter          metric.Meter
+	logger         *zap.Logger
+
+	// Otel API
```
Is this comment necessary here?

paivagustavo replied:
Not really 🙃

codeboten commented on obsreport/obsreport_receiver.go:
```diff
@@ -63,7 +83,70 @@ func NewReceiver(cfg ReceiverSettings) *Receiver {
+		},
 	}
+
+	rec.createOtelMetrics()
```
should this be behind a check for using otel metrics?

Aneurysm9 replied:
Or the check should be at the top of the method.  One of the two.

paivagustavo replied:
I've decided to add a check at the top.

codeboten reviewed (commented):
Just a couple of questions

Aneurysm9 commented on service/telemetry.go:
```diff
@@ -85,7 +81,7 @@ func init() {
 func registerInternalMetricFeatureGate(registry *featuregate.Registry) {
 	registry.MustRegister(featuregate.Gate{
-		ID:          useOtelForInternalMetricsfeatureGateID,
+		ID:          obsreportconfig.UseOtelForInternalMetricsfeatureGateID,
```
This registration should move to `obsreportconfig` if the ID constant is moving there as well.

paivagustavo replied:
done.

Aneurysm9 commented on obsreport/obsreport_receiver.go:
```diff
@@ -163,23 +246,11 @@ func (rec *Receiver) endOp(
-		case config.LogsDataType:
-			acceptedMeasure = obsmetrics.ReceiverAcceptedLogRecords
-			refusedMeasure = obsmetrics.ReceiverRefusedLogRecords
+		if rec.useOtelForMetrics {
```
Maybe hide this behind a `rec.record()` method that does the switching.

paivagustavo replied:
Created a `recordMetrics` to hide it and also to not be mistaken with the `if span.IsRecording() ` line below which is not refer a different type of "record"

paivagustavo commented:
Thanks for the reviews @bogdandrutu @codeboten @Aneurysm9, I think I've addressed all the comments at this point and it should be ready to be reviewed again.

codeboten reviewed (approved):
Thanks for addressing my comments 👍🏻

evantorrie commented on internal/obsreportconfig/obsreportconfig.go:
```diff
@@ -20,9 +20,30 @@ import (
+func RegisterInternalMetricFeatureGate(registry *featuregate.Registry) {
+	registry.MustRegister(featuregate.Gate{
+		ID:          UseOtelForInternalMetricsfeatureGateID,
+		Description: "controls whether the collector to uses OpenTelemetry for internal metrics",
```
```suggestion
		Description: "controls whether the collector uses OpenTelemetry for internal metrics",
```

bogdandrutu commented on obsreport/obsreport_receiver.go:
```diff
@@ -63,7 +83,74 @@ func NewReceiver(cfg ReceiverSettings) *Receiver {
 			tag.Upsert(obsmetrics.TagKeyTransport, cfg.Transport, tag.WithTTL(tag.TTLNoPropagation)),
 		},
 		tracer: cfg.ReceiverCreateSettings.TracerProvider.Tracer(cfg.ReceiverID.String()),
+		meter:  cfg.ReceiverCreateSettings.MeterProvider.Meter(cfg.ReceiverID.String()),
```
Not sure if this is the right name for the meter. Should we have a meter per factory type of per factory component instance?

paivagustavo replied:
Do you mean to use `receiver/otlp` for the `otlp` receiver's meter name instead of just `otlp`, to differentiate components that have the same name? 

If that is what you meant, I think that seems better. Should tracer name be like that as well?

bogdandrutu replied:
Nope, I mean to use `cfg.ReceiverID.Type()`, so if we have multiple instances of the `otlp` receiver we have same scope for this, and the instance  name is not in the scope name (probably should by an attribute).

Aneurysm9 replied:
I don't think `otlp` is an appropriate scope name.  Shouldn't the scope here be `go.opentelemetry.io/collector/obsreport` with attributes for `componentType=receiver` and `componentID=otlp` (or `componentID=otlp/foo` if a name exists on the receiver ID)?

bogdandrutu replied:
That is also a possibility, but you are still suggesting the same thing, that component name should not be part of the scope (which was before).

I think I like scope.name=`go.opentelemetry.io/collector/obsreport/receiver` and have a `component.id=otlp` (or `component.id=otlp/foo`) since we use `.` mostly as attribute name separator in otel (see resource attributes for example).

paivagustavo replied:
By the way, we will encounter some warns/missing metrics due to some otel-go bugs at the moment. 

Because the go sdk doesn't return a valid instrument for "duplicate" metric creation results in invalid instruments in otel-go ([tracking issue](https://github.com/open-telemetry/opentelemetry-go/issues/3229
)), only the first receiver would have valid instruments and subsequent calls to the `obsreport.NewReceiver()` will have noop metrics.

I would suggest us to just add the `component.id` (`cfg.ReceiverID.String()`) as suffix for now, adding a TODO with the tracking issue and remove that once that gets fixed. Since the prometheus exporter doesn't use the scope for nothing yet, it wouldn't change anything.

bogdandrutu replied:
I would not do that, since that will be a breaking change. But fix was release today.

paivagustavo replied:
Right, upgrading to the new released version fixed this.

paivagustavo commented on internal/obsreportconfig/obsreportconfig.go:
```diff
@@ -103,6 +116,26 @@ func allViews() []*view.View {
 }
 
+func receiverViews() []*view.View {
+	if featuregate.GetRegistry().IsEnabled(UseOtelForInternalMetricsfeatureGateID) {
```
kinda misc, but I've added featuregate check on the OC receiver views creation.

bogdandrutu replied:
why? I don't think we should code this way, the contract is to register these views only if `UseOtelForInternalMetricsfeatureGateID`, then we protect the high level function we don't have to protect any sub-function.

paivagustavo replied:
Because if we want to use `prometheus.Registry` as the bridge, we don't want OC to emit metrics that have being migrated to otel-go if the featuregate is enabled, but we would want all the others non-migrated oc metrics to be exported as normal, hence I just removed the `receiver` metrics for now.

Once all components of the core gets instrumented with otel-go, we can move this to be broader and not register any view at all when the featuregate is enabled, but for now I would suggest to just remove it component per component.

bogdandrutu commented on obsreport/obsreport_receiver.go:
```diff
@@ -63,7 +88,74 @@ func NewReceiver(cfg ReceiverSettings) *Receiver {
+	}
+
+	var err error
+	handleError := func(metricName string, err error) {
+		if err != nil {
+			rec.logger.Warn("failed to create otel instrument", zap.Error(err), zap.String("metric", metricName))
+		}
```
We need to return the error to the caller instead of logging. It should not be done in this PR, but we need to file an issue for this.

bogdandrutu reviewed (approved):
Just for my sanity, please create a separate PR for the version updates, maybe also include the 1.11 updates as well.
